### ISSUE TITLE
How to download TF modules from private Git repos during `terraform init` inside pre-build container

### ISSUE THREAD

**Issue by SRodi** [antonbabenko/pre-commit-terraform#513]
### What problem are you facing?

We need to scan a number of directories and subdirectories within our repo which represent our `live-environments`. The terraform templates defined in these directories source modules from private GitHub repositories, this was implemented to enable module versioning.

The pre-commit hook is failing with the following error:

```bash
│ Error: Failed to download module
│ 
│ Could not download module "my_module" (main.tf:50) source code from
│ "git::https://github.com/MyOrg/my-repo.git?ref=v1.14.3":
│ error downloading
│ 'https://github.com/MyOrg/my-repo.git?ref=v1.14.3':
│ /usr/bin/git exited with 128: Cloning into
│ '.terraform/modules/my_module'...
│ fatal: could not read Username for 'https://github.com': No such device
│ or address
│ 
```

### How could pre-commit-terraform help solve your problem?

Allow to automatically detect when modules are sourced from a private repo so that the token is inserted for the duration of the session.

The dynamic git configuration can be done on a local host as follows:

```bash
GITHUB_SERVER_URL=github.com // or any enterprise GH
git config --global url."https://oauth2:${TOKEN}@${GITHUB_SERVER_URL}".insteadOf "https://${GITHUB_SERVER_URL}"
```

### Change request

Allow to add a [PersonalAccessToken](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) to the container  so that `~/.gitconfig` would include the following entry

```bash
[url "https://oauth2:TOKEN@github.com"]
        isteadOf = https://github.com
```

I am not sure this is the correct approach, maybe `.netrc` is required to pull modules from private GitHub repos in this case? I can help implementing the change but it would be good to get some suggestions on how to tackle this issue.

Thank you

**Comment by MaxymVlasov** [comment]
Hi

Let's summarize: You are not able to use [pre-build images](https://github.com/antonbabenko/pre-commit-terraform/pkgs/container/pre-commit-terraform), because you need to provide not only PAT/etc inside the container but also configuration for git ~by ENV VARs?~.

Are you sure that you are not able in CI create the needed `.gitconfig`, and provide it via extra `--volume` to

`docker run -e "USERID=$(id -u):$(id -g)" -v $(pwd):/lint -w /lint pre-commit-terraform`?


Because we can add to `entrypoint.sh` something like

```bash
# Allow to ad-hock set up git configs and so on via ENV VARS
if [[ -z "${PRE_CONFIGURE_IMAGE_COMMANDS}" ]]; then
  eval "$PRE_CONFIGURE_IMAGE_COMMANDS"
fi
```

but I don't think that it is worth it.

**Comment by MaxymVlasov** [comment]
Different providers use a little bit different [settings](https://gist.github.com/StevenACoffman/866b06ed943394fbacb60a45db5982f2#how-to-go-get-private-repos-using-https-via-personal-access-tokens), and we somehow should not break public repos users, which have no token and provide support for over9999 possible cases of private auth.

So here should be or generic implementation like in the code block above, or a suggestion to extend/customize the image as you want, by building on top or from scratch and adding all that you need or by `docker run` arguments or directly to Dockerfile

**Comment by SRodi** [comment]
Thank you for the response @MaxymVlasov !

As suspected, to fix the issue with download of Terraform modules from private GitHub repositories, the [.netrc file](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html) can be used.

My problem was that the `.netrc` file was not placed in the correct directory within the container using the volume attachment.

Here is how I fixed it and what we do in our CI to download Terraform modules from private repos:

```bash
# create .netrc file on CI server
echo -e "machine github.ibm.com\n  login $GH_TOKEN" > ~/.netrc

# run pre-commit-terraform with docker
# adding volume for .netrc file
# note, .netrc needs to be in /root/ dir
docker run --rm -e "USERID=$(id -u):$(id -g)" -v ~/.netrc:/root/.netrc -v $(pwd):/lint -w /lint ghcr.io/antonbabenko/pre-commit-terraform:latest run -a
```

No changes are required to support this, thanks again.

**Comment by yermulnik** [comment]
@SRodi @MaxymVlasov Might this worth of being added into `Docker Usage` section for posterity?

**Comment by SRodi** [comment]
> @SRodi @MaxymVlasov Might this worth of being added into `Docker Usage` section for posterity?

@yermulnik I agree, I am happy to work on this and create a PR with the change if you want to reopen this issue and assign to me. Thanks

**Comment by yermulnik** [comment]
@SRodi This would be great if @MaxymVlasov has no objections.

**Comment by antonbabenko** [comment]
This issue has been resolved in version 1.79.0 :tada:
